lib/connection_tester.rb

```diff
@@ -0,0 +1,243 @@
+        ct.request(req_opts, res)
+
+        # test for the diaspora* version
+        ct.nodeinfo(res)
+
+      rescue Failure => e
+        case e
+        when AddressFailure, DNSFailure, NetFailure
+          res.reachable = false
+        when SSLFailure
+          res.reachable = true
+          res.ssl_status = false
+        when HTTPFailure
+          res.reachable = true
+          res.ssl_status = true if ct.uses_ssl?
+        when NodeInfoFailure
+          res.diaspora_version = ""
+        end
+
+        res.error = e
+      end
+
+      res.freeze
+    end
+
+    # assemble a hash of options for a Net::HTTP request.
+    # This also includes a :host key containing the hostname of the server
+    #
+    # @param [String] target URL
+    def request_opts_for_url(url)
+      uri = URI.parse(url)
+      raise AddressFailure, "invalid protocol: '#{uri.scheme.upcase}'" unless uri.is_a?(URI::HTTP) || uri.is_a?(URI::HTTPS)
+      ssl = (uri.scheme == 'https') ? true : false
+      port = ssl ? 443 : 80
+
+      REQUEST_OPTS.merge({
+        host: uri.host,
+        use_ssl: ssl,
+        port: port,
+        path: uri.path
+      })
+    rescue URI::InvalidURIError => e
+      raise AddressFailure, e.message
+    end
+  end
+
+  def initialize(hostname)
+    @hostname = hostname
+  end
+
+  # Perform the DNS query
+  #
+  # @param [Result]
+  # @return [Resolv::IPv4, Resolv::IPv6] the resolved IP address
+  def resolve(result)
+    res = Resolv::DNS.new
+    addr = res.getaddress(@hostname)
+    result.ip = addr.to_s
+  rescue Resolv::ResolvError, Resolv::ResolvTimeout => e
+    raise DNSFailure, "'#{@hostname}' - #{e.message}"
+  ensure
+    res.close
+  end
+
+  # Perform a HTTP GET request to determine the following information
+  #  * is the host reachable
+  #  * is port 80/443 open
+  #  * is the SSL certificate valid (only on HTTPS)
+  #  * does the server return a successful HTTP status code
+  # (can't do a HEAD request, since that's not a defined route in the app)
+  # follow up to 3 redirects by default
+  #
+  # @param [Hash] request options
+  # @param [Result]
+  # @return [Integer] HTTP status code
+  def request(opts, result, host=nil, max_redirects=3)
+    host ||= @hostname
+    raise HTTPFailure, "too many redirects on '#{host}'" if max_redirects == 0
+
+    port = opts.delete(:port)
+    path = opts.delete(:path)
+    @uses_ssl = opts[:use_ssl]
+    Net::HTTP.start(host, port, nil, nil, nil, opts) do |http|
+      start = Time.new
+      resp  = http.get(path.present? ? path : "/")
+      stop  = Time.new
+
+      result.status_code = Integer(resp.code)
+      result.rt          = ((stop-start) * 1000.0).to_i # milliseconds
+      case resp
+      when Net::HTTPSuccess
+        result.reachable      = true
+        result.ssl_status     = @uses_ssl
+      when Net::HTTPRedirection
+        # follow the redirect, decrement counter
+        req_opts = ConnectionTester.request_opts_for_url(resp['location'])
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.